compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -468,8 +468,8 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                 candidates.vec.push(AsyncClosureCandidate);
             }
             // Provide an impl, but only for suitable `fn` pointers.
-            ty::FnPtr(sig) => {
-                if sig.is_fn_trait_compatible() {
+            ty::FnPtr(sig_tys, hdr) => {
+                if sig_tys.with(hdr).is_fn_trait_compatible() {
```

Comment:
`is_fn_trait_compatible` should be implemented as a function on `FnHeader`

Reply:
Apparently most callsites either already create a `FnSig` immediately after, or don't have a header (e.g. `FnDef`)